src/v/coproc/tests/utils/supervisor.h

```diff
@@ -11,38 +11,65 @@
 #pragma once
 #include "coproc/supervisor.h"
 #include "coproc/tests/utils/coprocessor.h"
+#include "coproc/tests/utils/wasm_event_generator.h"
 #include "coproc/types.h"
-#include "model/record_batch_reader.h"
+#include "model/record.h"
 
 #include <seastar/core/circular_buffer.hh>
 #include <seastar/core/gate.hh>
+#include <seastar/core/map_reduce.hh>
 
 #include <absl/container/flat_hash_map.h>
-#include <absl/hash/hash.h>
 
 namespace coproc {
+using script_map_t = absl::btree_map<script_id, std::unique_ptr<coprocessor>>;
+
 // A super simplistic form of the javascript supervisor soley used for
-// the purposes of testing
+// the purposes of testing. The sharded instance of script_map_t contains the
+// same coprocessors across all cores. However due to the distribution of NTPs
+// across shards, record_batches from a particular input topic will always
+// arrive on the same shard.
 class supervisor final : public coproc::supervisor_service {
 public:
-    using copro_map
-      = absl::flat_hash_map<coproc::script_id, std::unique_ptr<coprocessor>>;
-
+    /// class constructor
+    ///
+    /// Reference to sharded state map is passed so unit tests can query it
+    /// poking and proding the internal state of the service for validity
     supervisor(
-      ss::scheduling_group sc,
-      ss::smp_service_group ssg,
-      ss::sharded<copro_map>& cp_map)
-      : supervisor_service(sc, ssg)
-      , _coprocessors(cp_map) {}
+      ss::scheduling_group, ss::smp_service_group, ss::sharded<script_map_t>&);
+
+    /// Called when new coprocessor(s) are to be deployed
+    ///
+    /// Simply instantiate the script and place it in the working set
+    ss::future<enable_copros_reply>
+    enable_coprocessors(enable_copros_request&&, rpc::streaming_context&) final;
+
+    /// Called when coprocessors(s) are to be removed
+    ///
+    /// Removes the script(s) from the map, deallocates their resources
+    ss::future<disable_copros_reply> disable_coprocessors(
+      disable_copros_request&&, rpc::streaming_context&) final;
 
-    ~supervisor() override { _gate.close().get(); }
```

Comment:
do you have a virtual destructor setup for your test harness class hierarchy (or is it needed)? i'm not sure if seastar/boost are creating instances of the fixture but it might be relevant

Reply:
For my fixtures no theres no virtual destructor hierarchy. I don't think its needed, I rely on redpanda_thread_fixture to properly shutdown the harness.